bitcoin/src/pow.rs

```diff
@@ -0,0 +1,1530 @@
+//! functions here are designed to be fast, by that we mean it is safe to use them to check headers.
+//!
+
+use core::convert::TryFrom;
+use core::fmt::{self, LowerHex, UpperHex};
+use core::ops::{Add, Div, Mul, Not, Rem, Shl, Shr, Sub};
+
+use crate::consensus::encode::{self, Decodable, Encodable};
+#[cfg(doc)]
+use crate::consensus::Params;
+use crate::hash_types::BlockHash;
+use crate::io::{self, Read, Write};
+
+/// Implements `From<$int>` for `$ty` where `$ty` is an unsigned integer type.
+macro_rules! impl_from_unsigned_type {
+    ($ty:ident, $int:ident) => {
+        impl From<$int> for $ty {
+            #[inline]
+            fn from(x: $int) -> $ty { $ty(U256::from(x)) }
+        }
+    };
+}
+
+/// Implements `TryFrom<$int>` for `$ty` where `$ty` is a signed integer type.
+macro_rules! impl_try_from_signed_type {
+    ($ty:ident, $int:ident) => {
+        impl TryFrom<$int> for $ty {
+            type Error = TryFromError;
+            #[inline]
+            fn try_from(x: $int) -> Result<Self, Self::Error> {
+                let x = U256::try_from(x)?;
+                Ok($ty(x))
+            }
+        }
+    };
+}
+
+/// Implements `Mul` and `Div`. Requires `u64::from($int)`.
+macro_rules! impl_div_mul {
+    ($ty:ident, $int:ident) => {
+        impl core::ops::Div<$int> for $ty {
+            type Output = $ty;
+            #[inline]
+            fn div(self, rhs: $int) -> Self {
+                let rhs = U256::from(u64::from(rhs));
+                $ty(self.0 / rhs)
+            }
+        }
+
+        impl Mul<$int> for $ty {
+            type Output = $ty;
+            #[inline]
+            fn mul(self, rhs: $int) -> Self { $ty(self.0.mul(rhs)) }
+        }
+
+        impl Mul<$ty> for $int {
+            type Output = $ty;
+            #[inline]
+            fn mul(self, rhs: $ty) -> $ty { $ty(self.mul(rhs.0)) }
+        }
+    };
+}
+
+// Implement traits and methods shared by `Target` and `Work`.
+macro_rules! do_impl {
+    ($ty:ident) => {
+        impl $ty {
+            /// Creates `Self` from a big-endian byte array.
+            #[inline]
+            pub fn from_be_bytes(bytes: [u8; 32]) -> $ty { $ty(U256::from_be_bytes(bytes)) }
+
+            /// Creates `Self` from a little-endian byte array.
+            #[inline]
+            pub fn from_le_bytes(bytes: [u8; 32]) -> $ty { $ty(U256::from_le_bytes(bytes)) }
+
+            /// Converts `self` to a big-endian byte array.
+            #[inline]
+            pub fn to_be_bytes(self) -> [u8; 32] { self.0.to_be_bytes() }
+
+            /// Converts `self` to a little-endian byte array.
+            #[inline]
+            pub fn to_le_bytes(self) -> [u8; 32] { self.0.to_le_bytes() }
+        }
+
+        impl_from_unsigned_type!($ty, u8);
+        impl_from_unsigned_type!($ty, u16);
+        impl_from_unsigned_type!($ty, u32);
+        impl_from_unsigned_type!($ty, u64);
+        impl_from_unsigned_type!($ty, u128);
```

Comment:
Oh, this will make it possible to call `42.into()` to get `Work` or `Target`. I was mostly against allowing that. I suggest an inherent method `from_raw<T: Into<u128>>(value: T)`. But still I'm not even sure if it's ever useful to work with such small numbers.